cmd/clusterctl/pkg/client/cluster/template.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"context"
+	"encoding/base64"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/repository"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TemplateClient has methods to work with templates stored in the cluster/out of the provider repository.
+type TemplateClient interface {
+	// GetFromConfigMap returns a workload cluster template from the given ConfigMap.
+	GetFromConfigMap(configMapNamespace, configMapName, configMapDataKey, targetNamespace string, listVariablesOnly bool) (repository.Template, error)
+
+	// GetFromURL returns a workload cluster template from the given URL.
+	GetFromURL(templateURL, targetNamespace string, listVariablesOnly bool) (repository.Template, error)
+}
+
+// templateClient implements TemplateClient.
+type templateClient struct {
+	proxy               Proxy
+	configClient        config.Client
+	gitHubClientFactory func(configVariablesClient config.VariablesClient) (*github.Client, error)
+}
+
+// ensure templateClient implements TemplateClient.
+var _ TemplateClient = &templateClient{}
+
+// newTemplateClient returns a templateClient.
+func newTemplateClient(proxy Proxy, configClient config.Client) *templateClient {
+	return &templateClient{
+		proxy:               proxy,
+		configClient:        configClient,
+		gitHubClientFactory: getGitHubClient,
+	}
+}
+
+func (t *templateClient) GetFromConfigMap(configMapNamespace, configMapName, configMapDataKey, targetNamespace string, listVariablesOnly bool) (repository.Template, error) {
```

Comment:
```suggestion
func (t *templateClient) GetFromConfigMap(namespace, name, dataKey, targetNamespace string, listVariablesOnly bool) (repository.Template, error) {
```

Simplify the signature by removing the redundant prefixes